packages/fxa-auth-server/lib/payments/stripe.ts

```diff
@@ -678,6 +678,134 @@ export class StripeHelper {
+   * previewInvoice return value. Coupon details are returned
+   * regardless of current validity (expiry, redeemability).
+   *
+   * Throws invalidPromoCode error if the promotion code does not
+   * exist for the provided priceId.
+   */
+  async retrieveCouponDetails({
+    country,
+    priceId,
+    promotionCode,
+  }: {
+    country: string;
+    priceId: string;
+    promotionCode: string;
+  }) {
+    try {
+      const stripePromotionCode = await this.retrievePromotionCodeForPlan(
+        promotionCode,
+        priceId
+      );
+
+      if (stripePromotionCode?.coupon.id) {
+        let invoice: Stripe.Invoice | undefined = undefined;
+        const stripeCoupon = await this.getCoupon(
+          stripePromotionCode?.coupon.id
+        );
+        try {
+          invoice = await this.previewInvoice({
+            country,
+            priceId,
+            promotionCode,
+          });
+        } catch {
+          // do nothing - the invoice may have thrown an invalidPromoCode error due to the code now being invalid and therefore no discount
+        }
+
+        const couponDetails: {
+          promotionCode: string;
+          type: string;
+          valid: boolean;
+          discountAmount?: number;
+          expired?: boolean;
+          maximallyRedeemed?: boolean;
+        } = {
+          promotionCode: promotionCode,
+          type: stripeCoupon.duration,
+          valid: false,
+        };
+
+        if (invoice?.discount && invoice?.total_discount_amounts) {
+          couponDetails.discountAmount =
+            invoice.total_discount_amounts[0].amount;
+        }
+
+        if (stripeCoupon.redeem_by) {
+          const expiry = new Date(stripeCoupon.redeem_by * 1000);
+          const now = new Date();
+          couponDetails.expired = now > expiry;
+        }
+
+        if (stripeCoupon.max_redemptions) {
+          couponDetails.maximallyRedeemed =
+            stripeCoupon.times_redeemed >= stripeCoupon.max_redemptions;
+        }
+
+        if (
+          couponDetails.discountAmount &&
+          !couponDetails.expired &&
+          !couponDetails.maximallyRedeemed &&
+          stripePromotionCode.active
+        ) {
+          couponDetails.valid = true;
+        }
+
+        return couponDetails;
+      } else {
+        throw error.invalidPromoCode(promotionCode);
+      }
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  /**
+   * Retrieves the stripe promotionCode object for a plan regardless of current validity.
+   */
+  async retrievePromotionCodeForPlan(
+    code: string,
+    priceId: string
+  ): Promise<Stripe.PromotionCode | undefined> {
+    const promotionCode = await this.findPromoCodeByCode(code, undefined);
+    if (!promotionCode) {
+      return;
+    }
+
+    const planContainsPromo = await this.checkPromotionCodeForPlan(
+      code,
+      priceId
+    );
+    if (!planContainsPromo) {
+      return;
+    }
+
+    return promotionCode;
+  }
```

Comment:
This function looks very similar to `findValidPromoCode`.  Can we combine some code?  Or short-circuit some logic?  Do we need to preview the invoice when we know we don't have a valid promo code?

Reply:
So I went down that rabbit hole and abstracted out as much shared code as I could but a few other things call that findValidPromoCode and it in turn calls a few other things that all focus around this idea that its doing some validity checks along the way and throwing invalidPromo errors along the way. For brevity, and without necessitating a big refactor, I think did as much code sharing as I could with preserving the invoice preview stuff in mind.